tools/dep_updaters/update-zlib.sh

```diff
@@ -1,23 +1,49 @@
 #!/bin/sh
 set -e
-# Shell script to update zlib in the source tree to a specific version
+# Shell script to update zlib in the source tree to the most recent version.
+# Zlib rarely creates tags or releases, so we use the latest commit on the main branch.
+# See: https://github.com/nodejs/node/pull/47417
 
 BASE_DIR=$(cd "$(dirname "$0")/../.." && pwd)
 DEPS_DIR="$BASE_DIR/deps"
 
-CURRENT_VERSION=$(grep "#define ZLIB_VERSION" "$DEPS_DIR/zlib/zlib.h" | sed -n "s/^.*VERSION \"\(.*\)\"/\1/p")
- 
-NEW_VERSION_ZLIB_H=$(curl -s "https://chromium.googlesource.com/chromium/src/+/refs/heads/main/third_party/zlib/zlib.h?format=TEXT" | base64 --decode)
+echo "Comparing latest upstream with current revision"
 
-NEW_VERSION=$(printf '%s' "$NEW_VERSION_ZLIB_H" | grep "#define ZLIB_VERSION" | sed -n "s/^.*VERSION \"\(.*\)\"/\1/p")
+# Revert zconf.h changes before checking diff 
+perl -i -pe 's|^//#include "chromeconf.h"|#include "chromeconf.h"|' $DEPS_DIR/zlib/zconf.h
+git stash
+git fetch https://chromium.googlesource.com/chromium/src/third_party/zlib.git HEAD
 
-echo "Comparing $NEW_VERSION with $CURRENT_VERSION"
+DIFF_TREE=$(
+  git diff stash@{0}:deps/zlib FETCH_HEAD -- zconf.h
+  git diff --exit-code --diff-filter=d HEAD:deps/zlib FETCH_HEAD -- ':!zconf.h' || echo 1
+)
 
-if [ "$NEW_VERSION" = "$CURRENT_VERSION" ]; then
+git stash drop
+
+if [ -z "$DIFF_TREE" ]; then
```

Comment:
I think we also want to skip if we end up with `echo 1` above, no?

Reply:
We want to skip only if we don't end up with `echo 1`. `git diff --exit-code` exit with code 0 and no output when there's no diff. That being said, we don't gain much from the `--exit-code` flag, I was suggesting it so we can use it directly as the test, but currently it doesn't give us much expect an additional line with the number 1 🤷‍♂️

Reply:
Yes, my point is that if it fails we should exit, but currently even if it fails we end up with a non empty string and continue.

Reply:
Here's how I would write it:

```sh
# Revert zconf.h changes before checking diff
perl -i -pe 's|^//#include "chromeconf.h"|#include "chromeconf.h"|' "$DEPS_DIR/zlib/zconf.h"
git stash -- "$DEPS_DIR/zlib/zconf.h"

git fetch https://chromium.googlesource.com/chromium/src/third_party/zlib.git HEAD

git diff --exit-code --diff-filter=d stash@{0}:deps/zlib FETCH_HEAD -- ':!zconf.h' || (\
  echo "Skipped because zlib is on the latest version." && \
  git stash drop &&\
  exit 0 \
)
git stash drop
```

The double `git stash drop` is a bit unsettling, but I feel that's a more straightforward way of thinking about what this code is doing (but maybe that's just me).

Reply:
Without `--exit-code` seems cleaner and easier to understand to me.

Reply:
> DIFF_TREE=$(git diff --diff-filter=d stash@{0}:deps/zlib FETCH_HEAD -- ':!zconf.h')3

With this check, we didn't get the diff of other possible changes to zconf.h
I think we still need two checks. One for checking the entire lib without zconf, and one for checking only the re-patched zconf
```
DIFF_TREE=$(
  git diff stash@{0}:deps/zlib FETCH_HEAD -- zconf.h
  git diff --diff-filter=d HEAD:deps/zlib FETCH_HEAD -- ':!zconf.h' || echo 1
)
```

Reply:
Shouldn’t it be `DIFF_TREE=$(git diff stash@{0}:deps/zlib FETCH_HEAD)`? `stash@{0}` is supposed to be exactly the same as upstream if there are no updates, no?

Reply:
Yes you are right